File: uis/client/mobile_app/lib/services/tenant_service.dart
import 'dart:convert';
import 'package:flutter/foundation.dart' show kIsWeb;
import 'package:shared_preferences/shared_preferences.dart';
import 'package:universal_html/html.dart' as html;
import '../config/app_config.dart';
import '../models/tenant.dart';
import '../services/api_service.dart';

/// TenantService - matches web app implementation
/// Handles tenant configuration fetching, storage, and retrieval
class TenantService {
  static const String _tenantStorageKey = 'tenant_config';
  static const String _tenantIdStorageKey = 'tenant_id';

  /// Fetch tenant configuration from API
  /// Matches web app: ${AppConfig.tenantEndpoint}/tenant/${tenantId}
  static Future<TenantConfig> fetchTenantConfig(String tenantId) async {
    try {
      final apiService = ApiService();
      print('Fetching tenant configuration for tenant ID: $tenantId');
      print('API Endpoint: ${AppConfig.tenantEndpoint}/tenant/$tenantId');
      final response = await apiService.get(
        '${AppConfig.tenantEndpoint}/tenant/$tenantId',
      );

      if (response.statusCode != 200) {
        throw Exception('Failed to fetch tenant config: ${response.statusCode}');
      }

      // Extract tenant data from response (matching web app structure)
      final serverResponse = response.data;
      final tenantData = serverResponse['tenant'] ?? serverResponse['data'] ?? serverResponse;
      
      if (tenantData == null) {
        throw Exception('No tenant data in API response');
      }

      // Transform tenant data (matching web app transformTenantData)
      final tenantConfig = TenantConfig.fromJson(tenantData as Map<String, dynamic>);
      
      // Save tenant ID to storage (matching web app)
      await saveTenantId(tenantId);
      
      return tenantConfig;
    } catch (error) {
      print('Error fetching tenant configuration: $error');
      rethrow;
    }
  }

  /// Save tenant configuration to storage
  /// Matches web app: localStorage.setItem('tenant_config', JSON.stringify(tenant))
  static Future<void> saveTenantConfig(TenantConfig tenant) async {
    try {
      final json = jsonEncode(tenant.toJson());
      if (kIsWeb) {
        html.window.localStorage[_tenantStorageKey] = json;
      } else {
        final prefs = await SharedPreferences.getInstance();
        await prefs.setString(_tenantStorageKey, json);
      }
    } catch (error) {
      print('Error saving tenant configuration: $error');
    }
  }

  /// Load tenant configuration from storage
  /// Matches web app: localStorage.getItem('tenant_config')
  static Future<TenantConfig?> loadTenantConfig() async {
    try {
      String? stored;
      if (kIsWeb) {
        stored = html.window.localStorage[_tenantStorageKey];
      } else {
        final prefs = await SharedPreferences.getInstance();
        stored = prefs.getString(_tenantStorageKey);
      }
      
      if (stored != null) {
        final json = jsonDecode(stored) as Map<String, dynamic>;
        return TenantConfig.fromJson(json);
      }
      return null;
    } catch (error) {
      print('Error loading tenant configuration: $error');
      return null;
    }
  }

  /// Clear tenant configuration from storage
  /// Matches web app: removes both 'tenant_config' and 'tenant_id'
  static Future<void> clearTenantConfig() async {
    try {
      if (kIsWeb) {
        html.window.localStorage.remove(_tenantStorageKey);
        html.window.localStorage.remove(_tenantIdStorageKey);
      } else {
        final prefs = await SharedPreferences.getInstance();
        await prefs.remove(_tenantStorageKey);
        await prefs.remove(_tenantIdStorageKey);
      }
    } catch (error) {
      print('Error clearing tenant configuration: $error');
    }
  }

  /// Save tenant ID to storage
  /// Matches web app: localStorage.setItem('tenant_id', tenantId)
  static Future<void> saveTenantId(String tenantId) async {
    try {
      if (kIsWeb) {
        html.window.localStorage[_tenantIdStorageKey] = tenantId;
      } else {
        final prefs = await SharedPreferences.getInstance();
        await prefs.setString(_tenantIdStorageKey, tenantId);
      }
    } catch (error) {
      print('Error saving tenant ID: $error');
    }
  }

  /// Get tenant ID from storage
  /// Matches web app: localStorage.getItem('tenant_id')
  static Future<String?> getTenantId() async {
    try {
      if (kIsWeb) {
        return html.window.localStorage[_tenantIdStorageKey];
      } else {
        final prefs = await SharedPreferences.getInstance();
        return prefs.getString(_tenantIdStorageKey);
      }
    } catch (error) {
      print('Error getting tenant ID: $error');
      return null;
    }
  }

  /// Validate tenant configuration
  static bool validateTenantConfig(TenantConfig tenant) {
    if (tenant.id.isEmpty || tenant.name.isEmpty || tenant.domain.isEmpty) {
      print('Invalid tenant configuration: missing required fields');
      return false;
    }

    // Note: Branding and features are checked in TenantConfig model
    return true;
  }
}

File: uis/client/web2/src/services/tenant_service.ts
import { AppConfig } from '../config/app_config';
import type { TenantConfig, TenantServerResponse } from '../models/tenant';
import { transformTenantData } from '../models/tenant';
import { apiService } from './api_service';

export class TenantService {
  private static readonly TENANT_STORAGE_KEY = 'tenant_config';
  private static readonly TENANT_ID_STORAGE_KEY = 'tenant_id';

  /**
   * Fetch tenant configuration from API
   */
  static async fetchTenantConfig(tenantId: string): Promise<TenantConfig> {
    try {
      const response = await apiService.get<TenantServerResponse>(`${AppConfig.tenantEndpoint}/tenant/${tenantId}`);
      
      if (response.status !== 200) {
        throw new Error(`Failed to fetch tenant config: ${response.statusText}`);
      }

      const serverResponse = response.data;
      console.log('[TenantService] Raw API response:', {
        tenantName: serverResponse.tenant?.name,
        branding: serverResponse.tenant?.branding,
      });
      
      const tenantConfig = transformTenantData(serverResponse.tenant);
      
      console.log('[TenantService] Transformed config:', {
        tenantName: tenantConfig.name,
        primaryColor: tenantConfig.branding.primary_color,
        secondaryColor: tenantConfig.branding.secondary_color,
      });
      
      // Save tenant ID to localStorage
      this.saveTenantId(tenantId);
      
      return tenantConfig;
    } catch (error) {
      console.error('Error fetching tenant configuration:', error);
      throw error;
    }
  }

  /**
   * Fetch tenant configuration by domain
   */
  static async fetchTenantByDomain(domain: string): Promise<TenantConfig> {
    try {
      const response = await apiService.get<TenantServerResponse>(`${AppConfig.tenantEndpoint}/tenant/by-domain`, {
        domain: domain,
      });
      
      if (response.status !== 200) {
        throw new Error(`Failed to fetch tenant by domain: ${response.statusText}`);
      }

      const serverResponse = response.data;
      const tenantConfig = transformTenantData(serverResponse.tenant);
      
      return tenantConfig;
    } catch (error) {
      console.error('Error fetching tenant by domain:', error);
      throw error;
    }
  }

  /**
   * Save tenant configuration to localStorage
   */
  static saveTenantConfig(tenant: TenantConfig): void {
    try {
      localStorage.setItem(this.TENANT_STORAGE_KEY, JSON.stringify(tenant));
    } catch (error) {
      console.error('Error saving tenant configuration:', error);
    }
  }

  /**
   * Load tenant configuration from localStorage
   */
  static loadTenantConfig(): TenantConfig | null {
    try {
      const stored = localStorage.getItem(this.TENANT_STORAGE_KEY);
      return stored ? JSON.parse(stored) : null;
    } catch (error) {
      console.error('Error loading tenant configuration:', error);
      return null;
    }
  }

  /**
   * Clear tenant configuration from localStorage
   */
  static clearTenantConfig(): void {
    try {
      localStorage.removeItem(this.TENANT_STORAGE_KEY);
      localStorage.removeItem(this.TENANT_ID_STORAGE_KEY);
    } catch (error) {
      console.error('Error clearing tenant configuration:', error);
    }
  }

  /**
   * Save tenant ID to localStorage
   */
  static saveTenantId(tenantId: string): void {
    try {
      localStorage.setItem(this.TENANT_ID_STORAGE_KEY, tenantId);
    } catch (error) {
      console.error('Error saving tenant ID:', error);
    }
  }

  /**
   * Get tenant ID from localStorage
   */
  static getTenantId(): string | null {
    try {
      return localStorage.getItem(this.TENANT_ID_STORAGE_KEY);
    } catch (error) {
      console.error('Error getting tenant ID:', error);
      return null;
    }
  }

  /**
   * Get tenant ID from URL query parameter
   */
  static getTenantIdFromURL(): string | null {
    const urlParams = new URLSearchParams(window.location.search);
    return urlParams.get('tenant') || urlParams.get('tenant_id') || urlParams.get('bank');
  }

  /**
   * Detect tenant from current hostname
   */
  static detectTenantFromHostname(): string {
    const hostname = window.location.hostname;
    
    // Remove common prefixes
    const cleanHostname = hostname
      .replace('www.', '')
      .replace('app.', '')
      .replace('portal.', '');

    // Extract subdomain if exists
    const parts = cleanHostname.split('.');
    if (parts.length > 2) {
      return parts[0]; // Return subdomain as tenant identifier
    }

    // Extract domain name as tenant identifier
    return parts[0];
  }

  /**
   * Validate tenant configuration
   */
  static validateTenantConfig(tenant: TenantConfig): boolean {
    if (!tenant.id || !tenant.name || !tenant.domain) {
      console.error('Invalid tenant configuration: missing required fields');
      return false;
    }

    if (!tenant.branding || !tenant.features) {
      console.error('Invalid tenant configuration: missing required sections');
      return false;
    }

    return true;
  }

  /**
   * Apply tenant branding to manifest.json for PWA
   */
  static updatePWAManifest(tenant: TenantConfig): void {
    const manifest = {
      name: tenant.displayName,
      short_name: tenant.name,
      description: `${tenant.displayName} - Digital Banking Platform`,
      theme_color: tenant.branding.primary_color,
      background_color: tenant.branding.backgroundColor,
      display: 'standalone',
      orientation: 'portrait',
      start_url: '/',
      icons: [
        {
          src: tenant.logo,
          sizes: '192x192',
          type: 'image/png',
        },
        {
          src: tenant.logo,
          sizes: '512x512',
          type: 'image/png',
        },
      ],
    };

    // Update manifest dynamically
    const manifestElement = document.querySelector('link[rel="manifest"]') as HTMLLinkElement;
    if (manifestElement) {
      const blob = new Blob([JSON.stringify(manifest)], { type: 'application/json' });
      const manifestURL = URL.createObjectURL(blob);
      manifestElement.href = manifestURL;
    }
  }

  /**
   * Get feature flag value for current tenant
   */
  static isFeatureEnabled(tenant: TenantConfig, feature: keyof TenantConfig['features']): boolean {
    if (!tenant.features) return false;
    return tenant.features[feature] as boolean;
  }

  /**
   * Get feature config for a specific feature
   */
  static getFeatureConfig(tenant: TenantConfig, feature: string): Record<string, unknown> | undefined {
    if (!tenant.features) return undefined;
    const configKey = `${feature}Config`;
    return tenant.features[configKey] as Record<string, unknown> | undefined;
  }

  /**
   * Check if tenant is active
   */
  static isTenantActive(tenant: TenantConfig): boolean {
    return tenant.status === 'active';
  }

  /**
   * Get auth realm from feature config
   */
  static getAuthRealm(tenant: TenantConfig): string | undefined {
    const authConfig = this.getFeatureConfig(tenant, 'auth');
    if (authConfig && typeof authConfig === 'object' && 'realm' in authConfig) {
      return authConfig.realm as string;
    }
    return undefined;
  }

  /**
   * Get accounts config
   */
  static getAccountsConfig(tenant: TenantConfig): unknown {
    const accountsConfig = this.getFeatureConfig(tenant, 'accounts');
    if (accountsConfig && typeof accountsConfig === 'object' && 'account' in accountsConfig) {
      return accountsConfig.account;
    }
    return accountsConfig;
  }
}
